kernel: introduce BOOL, TRUE, FALSE

fingolfin opened the pull request:
This is a resurrected version of PR #1883.

This adds `typedef UChar BOOL`; the resulting C type `BOOL` can be used to
indicate that the return type of a function is supposed to be a boolean.
Matching integer constants TRUE=1 and FALSE=0 are also added.

The reason we are not using type `bool` from stdbool.h and the constants
`true` and `false`, all available in C99, is the following: in C99, a pointer
is automatically converted to type bool if necessary. That means that if one
by accident writes `return False` instead of `return false` in a function with
return type `BOOL`, then no compiler warning is issued. Instead, a bug is
introduced, as `False` is a non-NULL pointer, and thus is cast to the `bool`
value `true`.

With this commit, we get warnings similar to the following if we mix up
TRUE/True or FALSE/False; while if using `bool`, we'd only get the latter two
warnings, but not the first one.

    src/objects.c:897:12: error: incompatible pointer to integer conversion
          returning 'Obj' (aka 'unsigned long **') from a function with result
          type 'BOOL' (aka 'unsigned char') [-Werror,-Wint-conversion]
        return False;
               ^~~~~

    src/objects.c:1185:39: error: pointer/integer type mismatch in conditional
          expression ('int' and 'Obj' (aka 'unsigned long **'))
          [-Werror,-Wconditional-type-mismatch]
        return (TNUM_OBJ(obj) == T_COMOBJ ? TRUE : False);
                                          ^ ~~~~   ~~~~~

    src/objects.c:1337:16: error: incompatible integer to pointer conversion
          returning 'int' from a function with result type 'Obj' (aka
          'unsigned long **') [-Werror,-Wint-conversion]
            return TRUE;
                   ^~~~

fingolfin commented on src/lists.h:
```diff
@@ -592,7 +592,7 @@ EXPORT_INLINE Int IS_TABLE_LIST(Obj list)
 *V  IsSSortListFuncs[<type>]  .  table of strictly sorted list test functions
 **
-**  'IS_SSORT_LIST' returns 2 if the list <list> is  a strictly  sorted  list
+**  'IS_SSORT_LIST' returns 1 if the list <list> is  a strictly  sorted  list
```
Note that I changed this to always return 0 or 1. Many (most?) of the functions installed in `IsSSortListFuncs` already return 0 or 1, but a few returned 0 or 2 as documented. Luckily, all callers (there are not many) only care whether the return value is zero or non-zero, so this didn't matter.

By changing the return value to `BOOL` and using `return TRUE` and `return FALSE`, I think this code got clearer.

DominikBernhardt commented on src/blister.h:
```diff
@@ -48,6 +48,7 @@ EXPORT_INLINE Int SIZE_PLEN_BLIST(Int plen)
     return sizeof(Obj) + (plen + BIPEB - 1) / BIPEB * sizeof(UInt);
 }
 
+
```
Why did you add this line? This is not consistently done in the GAP kernel. Do we want to have two empty lines before we habe a `/***[...]` line?

fingolfin replied:
I generally aim for two empty lines, but this is not codified anywhere (and I am not keen on trying to codify it explicitly or officially either, as I am afraid that will just lead down a road of endless bikeshedding). In this particular case, the functions before and after this all used two empty lines (later on in the file it is again inconsistent). I guess I inserted the empty line (two years ago...) out of reflex. I can drop it, but I think it doesn't hurt either.

DominikBernhardt replied:
My comment was more of a question than an intend to get you to change it back. I'm fine with it, I just wanted to know what the general philosophy was.

DominikBernhardt commented on src/calls.h:
```diff
@@ -271,7 +271,7 @@ EXPORT_INLINE ObjFunc_1ARGS HDLR_XARGS(Obj func)
 **
 *F  IS_FUNC( <obj> )  . . . . . . . . . . . . . check if object is a function
```
```suggestion
*F  IS_FUNC( <obj> )  . . . . . . . . . . . . . check if object <obj> is a function
```
Just a minor remark.

fingolfin replied:
Out of 3083 such comments ("short function description"), just 136 reference a variable in this way. 

I really think the comment is fine as it is (it conveys its meaning well, doesn't it?). In any case, if it is to be changed, I'd suggest to change it in more places, and in a PR dedicate to it.

DominikBernhardt replied:
I guess it is also a personal preference. Yes, you're right, this I probably clear enough and we can leave it as is.

DominikBernhardt commented on src/gasman.c:
```diff
@@ -772,10 +772,8 @@ void InitGlobalBag (
-    const Char *        cookie2,
-    UInt                byWhat )
+static BOOL
+IsLessGlobal(const Char * cookie1, const Char * cookie2, UInt byWhat)
```
A general question here: Above, you changed return values `0` and `1` to `FALSE` and `TRUE`, below in lines 783 and following (I can't mark that line as there is no change in it), you left the return values `0`, `1` and `-1`. Why did you change the return value of the function to BOOL then?

fingolfin replied:
This was a mistake, thank you for catching it. I'll restore the return type, and also rename the function (IMHO anything starting with `Is...` like this should return a BOOL)

fingolfin replied:
Actually, looking at the code in question closer, I think it simply is buggy (probably due to a past refactoring?) -- indeed, its caller clearly expects it to return a BOOL.

fingolfin replied:
I refactored the code in question in PR #3815

DominikBernhardt commented on src/hpc/traverse.c:
```diff
@@ -204,17 +204,17 @@ static void TraverseRegionFrom(TraversalState * traversal,
 }
 
-static int IsWritableOrImmutable(TraversalState * traversal, Obj obj)
+static BOOL IsWritableOrImmutable(TraversalState * traversal, Obj obj)
```
Same as in a comment above: This function still returns `0` and `1`. Maybe I don't understand something, but it seems to me like this should be `FALSE` and `TRUE`?

fingolfin replied:
I'll change this particular function to return TRUE/FALSE now; but I never intended for this PR to be the final word on BOOL/TRUE/FALSE, but rather the beginning; in particular, there are lots of places that still use 0/1 not FALSE/TRUE. And I think that's fine overall (after all, comparisons like `foo < bar` also return 0 and 1), though it won't hurt to change it to TRUE/FALSE over time. But it's not necessary to do it all at once.

DominikBernhardt replied:
Ok, fair enough. You're right, one could do that in future PRs.

DominikBernhardt commented on src/integer.h:
```diff
@@ -33,7 +33,7 @@ EXPORT_INLINE Int IS_LARGEINT(Obj obj)
 **  'IS_INT' returns 1 if 'obj' is either a large or an immediate integer
```
So the return value of the function now is `BOOL`, should this comment be updated to 
```suggestion
**  'IS_INT' returns TRUE if 'obj' is either a large or an immediate integer
```

DominikBernhardt replied:
There are a bunch of places where this is the case. I don't point them all out now because I first would like to hear your opinion on this.

fingolfin replied:
I am reluctant to touch them because to preserve the line width limit one has to rewrap most of these comments, and it didn't seem worth the pain. But I wouldn't object if somebody else wanted to do this in another PR once this one is merged  :-)

DominikBernhardt commented on src/lists.c:
```diff
@@ -77,14 +77,14 @@ static Int IsListObject(Obj obj)
 static Obj SetIsSmallList;
 
-static Int IsSmallListObject(Obj obj)
+static BOOL IsSmallListObject(Obj obj)
```
Same as above, some return values are still `1` and `0`.

fingolfin replied:
I changed this now, too, but as explained above, I don't really want to change all such places in this PR; I'd rather leave this to a follow up PR (or multiple)

DominikBernhardt commented on src/objects.h:
```diff
@@ -562,8 +562,8 @@ void CheckedMakeImmutable(Obj obj);
-extern Int (*IsMutableObjFuncs[LAST_REAL_TNUM+1]) ( Obj obj );
-EXPORT_INLINE Int IS_MUTABLE_OBJ(Obj obj)
+extern BOOL (*IsMutableObjFuncs[LAST_REAL_TNUM + 1])(Obj obj);
+EXPORT_INLINE BOOL IS_MUTABLE_OBJ(Obj obj)
```
This also returns `0`.

fingolfin replied:
See above

fingolfin commented:
Turns out the semigroups package uses #define BOOL... this breaks all. So either we rename BOOL, or semigroups needs to change

DominikBernhardt commented:
Oh bummer. We could replace it with something like `GAPBOOL` or even `GAP_BOOL`, although both are longer and look a bit clumsier..

DominikBernhardt commented:
@fingolfin Am I stupid, but when I grep for `BOOL` in semigroups, I only find `BOOL_MAT`?

fingolfin commented:
It seems to be in libsemigroup, see Travis build logs

fingolfin commented:
I've added a workaround for the conflict with Semigroups (namely, I added `#undef BOOL` to `common.h`), and also submitted https://github.com/gap-packages/Semigroups/pull/635

DominikBernhardt commented on src/system.c:
```diff
@@ -42,6 +42,9 @@
 #include <sys/stat.h>
 
 #ifdef SYS_IS_DARWIN
+// Workaround for the fact that TRUE / FALSE area also
```
```suggestion
// Workaround for the fact that TRUE / FALSE are also
```

fingolfin replied:
Oops, fixed, thanks

DominikBernhardt commented:
Thanks! Now that [gap-packages/Semigroups#635](https://github.com/gap-packages/Semigroups/pull/635) is merged, are you planning to remove the workaround in this PR?

fingolfin commented:
I can only remove the workaround once a new Semigroups release has happened *and* the package distribution has it. Anyway, I think we might want to keep the workaround for some time (perhaps 1-2 full release cycles), as it allows people to keep using their existing semigroups installation.

fingolfin commented:
One thing we did not discuss is the name of that type. One might argue that the name `Bool` would fit better than `BOOL` (more similar to `Int`, `Char`, `Obj`, `Bag`, ...).  I think it would also conflict with fewer external sources! The reason I chose `BOOL` was that it seemed to fit better with `TRUE` and `FALSE`. And those were chosen because `True` and `False` already exist, of type `Obj` . I was concerned that it might be confusing that `True` is not of type `Bool` but `TRUE` is. While `TRUE` is `BOOL` but `True` is `Obj` seems more plausible.

fingolfin commented:
@wucas `bool` is a type provided by C++, and also by C99 (if one includes `stdbool.h`), see e.g. <https://en.cppreference.com/w/c/types/boolean>. The nice thing about using `bool` would be that it is standardized, and well-established in C++ code bases. The drawback for our purposes however is that a lot of things are automatically treated as `bool` by the C compiler. This is a feature in general, but undesirable for us: The feature is meant to ensure that things like `if (ptr) { ... }` just "work": `if(expr)` expects an expression evaluating to `bool`; but a `ptr` is not a `bool`; but in C, there is the implicit assumption that a null pointer is `false` and anything else is `true`, to ensure this kind of check (which is ubiquitous in C code) keeps working as expected. But for us, that's a problem, because we have global constants `True` and `False`, which are pointers to GAP objects representing `True` and `False`. So it would be quite easy to accidentally mixup `True` and `False` with `true` and `false`; and due to the above "feature" of C, we can only catch a subset of these wrong uses, even with the strictest compiler warnings enabled.

This is why this PR does not use the standard type `bool`, but rather invents its own type, `BOOL`.

wucas commented:
>@wucas `bool` is a type provided by C++, and also by C99 (if one includes `stdbool.h`), see e.g. <https://en.cppreference.com/w/c/types/boolean>. The nice thing about using `bool` would be that it is standardized, and well-established in C++ code bases. The drawback for our purposes however is that a lot of things are automatically treated as `bool` by the C compiler. This is a feature in general, but undesirable for us: The feature is meant to ensure that things like `if (ptr) { ... }` just "work": `if(expr)` expects an expression evaluating to `bool`; but a `ptr` is not a `bool`; but in C, there is the implicit assumption that a null pointer is `false` and anything else is `true`, to ensure this kind of check (which is ubiquitous in C code) keeps working as expected. But for us, that's a problem, because we have global constants `True` and `False`, which are pointers to GAP objects representing `True` and `False`. So it would be quite easy to accidentally mixup `True` and `False` with `true` and `false`; and due to the above "feature" of C, we can only catch a subset of these wrong uses, even with the strictest compiler warnings enabled. ...

@fingolfin thank you very much for your answer I think I understand it now. Also my approval still stands.

wilfwilson commented:
I think `BOOL` is probably the best choice at the moment but don’t care that much.

ChrisJefferson commented:
I'm happy with this. Seeing as we can't compile-time check if BOOL, TRUE and FALSE are used consistently/correctly, I'm not positive how much gain we will get, but I don't think it makes anything worse.

fingolfin commented:
@ChrisJefferson For me the main thing to gain is that it now is obvious that certain functions expect or return a boolean value, and that certain variables only take on boolean values (see also the recent question by @wucas on PR #3823); I also don't have to fret whether to make boolean variables be of type int, Int, UInt, char, ... and don't have to fret over code where these are all mixed. So for me, the main advantage is that it actually simplifies for me, mentally. Of course that may be highly subjective; but as long as nobody here feels actively *hindered* by this, I am hoping it'll be net positive, albeit indeed a small one.

fingolfin commented:
That said, I wish somebody would make an C variant where there is a type `bool` and it is a non-integral type, with no implicit conversion between it and `int`, pointers, whatever. No other change to the language. Then I could adapt all my C code to be compliant with it, by turning things like `if (ptr)` into `if (ptr != 0)` etc. If one does this right, the resulting code would still be valid C code and compile warning free, but I'd be sure I didn't accidentally mix up booleans and non-booleans.

I think there are a few other similar minor changes one could make; restrictions to the language where the resulting code still would be valid C code, and thus usable in lots of places, but one removes a "feature" that perhaps once seemed convenient, but now is burdensome. Unfortunately, people who look into such things apparently can't resist the urge to start making bigger and bigger changes until they arrive at a completely different language, like D. Ah well :-)
